File: 2018/07-jul/7-29-overloading/swift/Animal.swift
/*
 Struct representing an animal of a specific species
 - Author: Andrew Jarombek
 - Date: 7/28/2018
 */

import Foundation

struct Animal {
    
    let name: String
    let species: AnimalSpecies
    let description: String?
    
    /**
     Constructor for an Animal of a specific species with a potentially empty description
     - parameters:
     - name: A string representing a name given to the animal
     - species: The specific species of the animal
     - description - A string describing the animal, which can be ommitted
     */
    init(name: String, species: AnimalSpecies, description: String?) {
        self.name = name
        self.species = species
        self.description = description
    }
    
    /**
     Constructor for an Animal of a specific species
     - parameters:
     - name: A string representing a name given to the animal
     - species: The specific species of the animal
     */
    init(name: String, species: AnimalSpecies) {
        self.init(name: name, species: species, description: nil)
    }
}

extension Animal: Equatable {
    
    /**
     Extend the Animal class and implement the equality operator.  Checks if two animals can be
     determined as equal.  Animals are equal if their name, species, and description fields are
     all the same
     - parameters:
     - lhs: the first animal to test for equality
     - rhs: the second animal to test for equality
     */
    static func ==(lhs: Animal, rhs: Animal) -> Bool {
        return lhs.name == rhs.name && lhs.species == rhs.species && lhs.description == rhs.description
    }
}

enum AnimalSpecies {
    case Deer, Groundhog, Rabbit, Raccoon, Squirrel, Chipmunk, Crow, Cardinal
}

File: 2018/07-jul/7-29-overloading/php/Animal.php
<?php
/**
 * Class representing an animal of a specific species
 * User: Andrew Jarombek
 * Date: 7/28/2018
 */

class Animal
{
    private $name;
    private $species;
    private $description;

    /**
     * Animal constructor that takes in an optional description.
     * @param $name - a name given to the animal
     * @param $species - the specific species of the animal
     * @param string $description - an optional description of the animal (defaults to an empty string)
     */
    public function __construct($name, $species, $description="")
    {
        $this->name = $name;
        $this->species = $species;
        $this->description = $description;
    }
}